### DISCUSSION TITLE
Possibility of using libCEED for mortar contact methods

### DISCUSSION THREAD

**Discussion by lindsayad** [CEED/libCEED#1659] [Q&A] [answered]
For a Lagrange multiplier based contact method in MOOSE we define the Lagrange multiplier degrees of freedom on lower dimensional elements along the contact surface. So my first question is whether it’s possible to have multiple element restrictions for a given L-vector? The L-vector here holding both displacement degrees of freedom living on the higher dimensional elements and the lower dimensional element LM degrees of freedom

**Comment by knepley** [comment]
Yes, because in libCEED it is completely algebraic. Now someone has to generate this restriction from a bunch of crazy configurations, but that is the user's problem.

**Comment by jrwrigh** [comment, marked as answer]
> whether it’s possible to have multiple element restrictions for a given L-vector?

I'm not sure I follow the premise, but yes, it's possible to have multiple element restrictions defined for the same L-vector. This is what we do for boundary ~~conditions~~ integrals, and what we'd do for multi-topology meshes. Each unique grouping of elements (boundary elements, tets, wedges, etc.) has it's own `CeedElemRestriction` object describing the mapping between the L-vector entries and each element.

**Reply by lindsayad** [reply]
Great, this answers my question!

**Reply by jedbrown** [reply]
Cc: @zatkins-dev who has been doing work with a few contact formulations in Ratel (based on libCEED and PETSc). I'd also be happy to join if you want to go deeper on the topic of contact with libCEED/GPUs.

**Reply by zatkins-dev** [reply]
I would definitely be interested in helping to build some utilities to make that translation easier for libCEED users. I want to get mesh-to-mesh contact working in Ratel with libCEED and PETSc, so we would also need those.

**Reply by lindsayad** [reply]
@knepley's answer made me think that there is no development required in libCEED itself to make this work, since libCEED just does algebra. I assumed then that it's the job of the finite element library to determine the basis function evaluations that would be a function of mesh motion and geometric projections

**Reply by jedbrown** [reply]
A common situation is that you need to evaluate one element at projections of quadrature points from a different element. That can be done with the AtPoints interface, though you would be responsible for matching up those points and determining which element to evaluate in. It might be that libCEED should do something to make this easier, but AtPoints may be sufficient. We may want to set up a 2D patch test or ironing problem (i.e., topologically 1D contact surface) to work through this in a small/simple environment.

**Reply by lindsayad** [reply]
Yes that all sounds good. We have both patch and ironing verification tests for making sure a future implementation is correct
